cardano-api/src/Cardano/Api/Typed.hs

```diff
@@ -1279,35 +1272,102 @@ makeShelleyBootstrapWitness nw (ShelleyTxBody txbody _) (ByronSigningKey sk) =
 
 data ShelleyWitnessSigningKey =
        WitnessPaymentKey         (SigningKey PaymentKey)
+     | WitnessPaymentExtendedKey (SigningKey PaymentExtendedKey)
      | WitnessStakeKey           (SigningKey StakeKey)
      | WitnessStakePoolKey       (SigningKey StakePoolKey)
      | WitnessGenesisDelegateKey (SigningKey GenesisDelegateKey)
      | WitnessGenesisUTxOKey     (SigningKey GenesisUTxOKey)
 
+
 makeShelleyKeyWitness :: TxBody Shelley
                       -> ShelleyWitnessSigningKey
                       -> Witness Shelley
-makeShelleyKeyWitness (ShelleyTxBody txbody _) wsk =
-    ShelleyKeyWitness $
-      Shelley.WitVKey vk signature
-  where
-    sk = toShelleySignKey wsk
-    vk = Shelley.VKey (Crypto.deriveVerKeyDSIGN sk)
-
-    signature :: Shelley.SignedDSIGN ShelleyCrypto
-                  (Shelley.Hash ShelleyCrypto (Shelley.TxBody ShelleyCrypto))
-    signature = Crypto.signedDSIGN () txhash sk
+makeShelleyKeyWitness (ShelleyTxBody txbody _) =
+    let txhash :: Shelley.Hash ShelleyCrypto (Shelley.TxBody ShelleyCrypto)
+        txhash = Crypto.hash txbody
 
-    txhash :: Shelley.Hash ShelleyCrypto (Shelley.TxBody ShelleyCrypto)
-    txhash = Crypto.hash txbody
+        -- To allow sharing of the txhash computation across many signatures we
+        -- define and share the txhash outside the lambda for the signing key:
+     in \wsk ->
+        let sk        = toShelleySigningKey wsk
+            vk        = getShelleyKeyWitnessVerificationKey sk
+            signature = makeShelleyKeyWitnessSignature txhash sk
+         in ShelleyKeyWitness $
+              Shelley.WitVKey vk signature
```

Comment:
I'm not sure I understand this change.

Does this allow us to avoid constantly recomputing the `txhash` value when mapping a partially applied `makeShelleyKeyWitness` over many `ShelleyWitnessSigningKey`s, for example?

Reply:
Yes, exactly that.